File: pressf/stats.py
"""Statistics for trusting numbers: confidence intervals, judge's accuracy, sampling.

Everything is pure functions with no Project dependencies to test directly."""

from __future__ import annotations

import math
import random
from dataclasses import dataclass
from typing import Sequence, TypeVar

T = TypeVar("T")


def wilson_interval(successes: int, n: int, z: float = 1.96) -> tuple[float, float]:
    """Wilson confidence interval for the proportion (default 95%).

    Wilson, and not the normal approximation - it is correct both at small n and at the edges 0/1,
    where the usual «p ± z·se» goes beyond [0,1] or collapses to zero."""
    if n <= 0:
        return (0.0, 0.0)
    p = successes / n
    denom = 1.0 + z * z / n
    center = (p + z * z / (2 * n)) / denom
    margin = (z * math.sqrt(p * (1 - p) / n + z * z / (4 * n * n))) / denom
    return (max(0.0, center - margin), min(1.0, center + margin))


def sign_test_p(successes: int, n: int) -> float:
    """Accurate two-way sign test for pairwise comparisons.

    Draws do not fall into n: the test answers whether one version could win so many times
    happen by chance with equal versions."""
    if n <= 0:
        return 1.0
    observed = math.comb(n, successes)
    numerator = sum(math.comb(n, i) for i in range(n + 1) if math.comb(n, i) <= observed)
    return min(1.0, numerator / (2**n))


@dataclass(frozen=True)
class PairwiseSummary:
    a_wins: int
    b_wins: int
    ties: int
    decided: int
    b_win_rate: float | None
    ci_low: float
    ci_high: float
    p_value: float | None
    left_pick_rate: float | None

    @property
    def decision(self) -> str:
        """Conservative recommendation for version B release."""
        if not self.decided:
            return "There are not enough solved pairs for withdrawal."
        if self.ci_low > 0.5:
            return "Version B is better: lower bound 95% CI greater than 50%; can be released."
        if self.ci_high < 0.5:
            return "Version B is worse: the upper limit of the 95% CI is below 50%; don't let her out."
        return "We can't say for sure that version B is better; add steam for marking."


def pairwise_summary(annotations: Sequence[object]) -> PairwiseSummary:
    """Aggregate effective PairwiseAnnotation without depending on Project."""
    a_wins = b_wins = ties = left_picks = 0
    for ann in annotations:
        winner = getattr(ann, "winner", None) or getattr(ann, "choice", None)
        if winner == "a":
            a_wins += 1
        elif winner == "b":
            b_wins += 1
        elif winner == "tie":
            ties += 1
        shown_left = getattr(ann, "shown_left", None)
        if winner in ("a", "b") and shown_left:
            left_picks += int(winner == shown_left)
    decided = a_wins + b_wins
    ci_low, ci_high = wilson_interval(b_wins, decided)
    return PairwiseSummary(
        a_wins=a_wins,
        b_wins=b_wins,
        ties=ties,
        decided=decided,
        b_win_rate=b_wins / decided if decided else None,
        ci_low=ci_low,
        ci_high=ci_high,
        p_value=sign_test_p(b_wins, decided) if decided else None,
        left_pick_rate=left_picks / decided if decided else None,
    )


def flag_precision_recall(pairs: Sequence[tuple[str, str]]) -> dict[str, float | int]:
    """Judge's accuracy on class «problem» (f) against a human reference.

    pairs - list (judge's recommendation, person's tag), tags in {p, f}.
    Positive class - f (found a problem). precision = of those marked by the judge as f,
    how much f is real per person; recall = from real f, how many the judge caught."""
    tp = fp = fn = 0
    for judge, human in pairs:
        if judge == "f" and human == "f":
            tp += 1
        elif judge == "f" and human == "p":
            fp += 1
        elif judge == "p" and human == "f":
            fn += 1
    precision = tp / (tp + fp) if (tp + fp) else 0.0
    recall = tp / (tp + fn) if (tp + fn) else 0.0
    f1 = 2 * precision * recall / (precision + recall) if (precision + recall) else 0.0
    return {"tp": tp, "fp": fp, "fn": fn, "precision": precision, "recall": recall, "f1": f1}


def per_category_agreement(rows: Sequence[tuple[str, str, str]]) -> dict[str, tuple[int, float]]:
    """Agreement between a judge and a person in terms of verdict categories.

    rows — (judge category, judge recommendation, person tag). Returns for each
    categories (how many are marked, percentage of agreement). Shows what types of errors
    the judge systematically disagrees with the person."""
    agg: dict[str, list[int]] = {}
    for category, judge, human in rows:
        if human not in ("p", "f"):
            continue
        total, agreed = agg.setdefault(category, [0, 0])
        agg[category] = [total + 1, agreed + int(judge == human)]
    return {cat: (total, agreed / total if total else 0.0) for cat, (total, agreed) in agg.items()}


def cohen_kappa(pairs: Sequence[tuple[str, str]]) -> float:
    """Cohen's kappa for the agreement of two markers on common examples.

    pairs - labels (annotator A, annotator B) for the same examples. Kappa cleans up
    agreement by chance: 1.0 - perfect, 0 - at the level of chance, <0 - worse than chance."""
    n = len(pairs)
    if n == 0:
        return 0.0
    labels = sorted({lbl for pair in pairs for lbl in pair})
    observed = sum(1 for a, b in pairs if a == b) / n
    a_freq = {lbl: sum(1 for a, _ in pairs if a == lbl) / n for lbl in labels}
    b_freq = {lbl: sum(1 for _, b in pairs if b == lbl) / n for lbl in labels}
    expected = sum(a_freq[lbl] * b_freq[lbl] for lbl in labels)
    if expected >= 1.0:
        return 1.0  #all marks match and are the same - complete agreement
    return (observed - expected) / (1 - expected)


def inter_annotator_kappa(
    per_annotator: dict[str, dict[str, str]]
) -> list[tuple[str, str, int, float]]:
    """Pairwise kappa between all markers who have common examples.

    per_annotator - {name: {example_id: label}}. Returns a list
    (annotator A, annotator B, number of common examples, kappa), only for pairs with overlap."""
    names = sorted(per_annotator)
    out: list[tuple[str, str, int, float]] = []
    for i in range(len(names)):
        for j in range(i + 1, len(names)):
            a, b = names[i], names[j]
            common = set(per_annotator[a]) & set(per_annotator[b])
            if not common:
                continue
            pairs = [(per_annotator[a][eid], per_annotator[b][eid]) for eid in sorted(common)]
            out.append((a, b, len(common), cohen_kappa(pairs)))
    return out


def seeded_sample(items: Sequence[T], n: int, seed: int = 0) -> list[T]:
    """Deterministic random sampling of n elements (for cheap representative testing).

    Preserves the original order of the selected ones - the run is predictable and reproducible by seed."""
    if n >= len(items):
        return list(items)
    if n <= 0:
        return []
    rng = random.Random(seed)
    idx = sorted(rng.sample(range(len(items)), n))
    return [items[i] for i in idx]

File: app/src/shared/stats.ts
import type { Annotation, Label, Verdict } from "../main/types";

// TypeScript port of pressf/stats.py and pressf/scoring.py so the desktop app can
// show the same trust numbers as out/report.md without spawning the CLI.

export function wilsonInterval(successes: number, n: number, z = 1.96): [number, number] {
  if (n <= 0) return [0, 0];
  const p = successes / n;
  const denom = 1 + (z * z) / n;
  const center = (p + (z * z) / (2 * n)) / denom;
  const margin = (z * Math.sqrt((p * (1 - p)) / n + (z * z) / (4 * n * n))) / denom;
  return [Math.max(0, center - margin), Math.min(1, center + margin)];
}

export type FlagQuality = {
  tp: number;
  fp: number;
  fn: number;
  precision: number;
  recall: number;
  f1: number;
};

export function flagPrecisionRecall(pairs: Array<[string, string]>): FlagQuality {
  let tp = 0;
  let fp = 0;
  let fn = 0;
  for (const [judge, human] of pairs) {
    if (judge === "f" && human === "f") tp += 1;
    else if (judge === "f" && human === "p") fp += 1;
    else if (judge === "p" && human === "f") fn += 1;
  }
  const precision = tp + fp ? tp / (tp + fp) : 0;
  const recall = tp + fn ? tp / (tp + fn) : 0;
  const f1 = precision + recall ? (2 * precision * recall) / (precision + recall) : 0;
  return { tp, fp, fn, precision, recall, f1 };
}

export function perCategoryAgreement(rows: Array<[string, string, string]>): Map<string, { total: number; agreement: number }> {
  const agg = new Map<string, { total: number; agreed: number }>();
  for (const [category, judge, human] of rows) {
    if (human !== "p" && human !== "f") continue;
    const bucket = agg.get(category) ?? { total: 0, agreed: 0 };
    bucket.total += 1;
    bucket.agreed += Number(judge === human);
    agg.set(category, bucket);
  }
  return new Map([...agg.entries()].map(([category, { total, agreed }]) => [category, { total, agreement: total ? agreed / total : 0 }]));
}

export function cohenKappa(pairs: Array<[string, string]>): number {
  const n = pairs.length;
  if (n === 0) return 0;
  const labels = [...new Set(pairs.flat())];
  const observed = pairs.filter(([a, b]) => a === b).length / n;
  const expected = labels.reduce((sum, label) => {
    const aFreq = pairs.filter(([a]) => a === label).length / n;
    const bFreq = pairs.filter(([, b]) => b === label).length / n;
    return sum + aFreq * bFreq;
  }, 0);
  if (expected >= 1) return 1; // everyone used one identical label — full agreement
  return (observed - expected) / (1 - expected);
}

export type AnnotatorKappa = { a: string; b: string; common: number; kappa: number };

export function interAnnotatorKappa(perAnnotator: Map<string, Map<string, Label>>): AnnotatorKappa[] {
  const names = [...perAnnotator.keys()].sort();
  const out: AnnotatorKappa[] = [];
  for (let i = 0; i < names.length; i += 1) {
    for (let j = i + 1; j < names.length; j += 1) {
      const a = perAnnotator.get(names[i])!;
      const b = perAnnotator.get(names[j])!;
      const common = [...a.keys()].filter((id) => b.has(id)).sort();
      if (!common.length) continue;
      const pairs = common.map((id) => [a.get(id)!, b.get(id)!] as [string, string]);
      out.push({ a: names[i], b: names[j], common: common.length, kappa: cohenKappa(pairs) });
    }
  }
  return out;
}

// Effective (last non-undone) label per example, split by annotator — the input
// interAnnotatorKappa needs. Mirrors effective_annotations_by_annotator() in config.py.
export function effectiveByAnnotator(annotations: Annotation[]): Map<string, Map<string, Label>> {
  const out = new Map<string, Map<string, Label>>();
  for (const ann of annotations) {
    const name = ann.annotator?.trim() || "unknown";
    const byExample = out.get(name) ?? new Map<string, Label>();
    if (ann.undone) byExample.delete(ann.example_id);
    else byExample.set(ann.example_id, ann.label);
    out.set(name, byExample);
  }
  for (const [name, byExample] of out) if (!byExample.size) out.delete(name);
  return out;
}

export type FaithfulnessScore = {
  faithfulness: number;
  passed: number;
  failed: number;
  n: number;
  source: "human" | "judge" | "none";
};

export function faithfulnessFromLabels(labels: string[]): FaithfulnessScore {
  const passed = labels.filter((label) => label === "p").length;
  const failed = labels.filter((label) => label === "f").length;
  const n = passed + failed;
  return { faithfulness: n ? passed / n : 0, passed, failed, n, source: "none" };
}

// Port of scoring.py score_project: human labels are the reference, judge verdicts
// only stand in while nothing is reviewed yet.
export function faithfulnessScore(effective: Record<string, Annotation>, verdicts: Record<string, Verdict>): FaithfulnessScore {
  const human = Object.values(effective).map((ann) => ann.label).filter((label) => label === "p" || label === "f");
  if (human.length) return { ...faithfulnessFromLabels(human), source: "human" };
  const judge = Object.values(verdicts).map((verdict) => verdict.recommendation).filter((label) => label === "p" || label === "f");
  if (judge.length) return { ...faithfulnessFromLabels(judge), source: "judge" };
  return { faithfulness: 0, passed: 0, failed: 0, n: 0, source: "none" };
}

// Exact two-sided sign test: probability of an outcome at least as extreme as k
// successes out of n fair coin flips. Log-space so n in the thousands doesn't underflow.
export function signTestP(k: number, n: number): number {
  if (n <= 0) return 1;
  const logC: number[] = [0];
  for (let i = 1; i <= n; i += 1) logC.push(logC[i - 1] + Math.log(n - i + 1) - Math.log(i));
  const logHalfN = n * Math.log(0.5);
  const logPmf = (i: number) => logC[i] + logHalfN;
  const observed = logPmf(k);
  let p = 0;
  for (let i = 0; i <= n; i += 1) {
    if (logPmf(i) <= observed + 1e-9) p += Math.exp(logPmf(i));
  }
  return Math.min(1, p);
}

export type PairwiseSummary = {
  aWins: number;
  bWins: number;
  ties: number;
  decided: number;
  winRate: number | null;      // share of decided pairs won by B (the new version)
  ci: [number, number];        // Wilson 95% CI on winRate
  pValue: number | null;       // sign test on decided pairs
  leftShare: number | null;    // how often the reviewer picked whichever answer was shown left
};

export function pairwiseSummary(effective: Record<string, { winner?: string; choice?: string; shown_left?: string }>): PairwiseSummary {
  let aWins = 0;
  let bWins = 0;
  let ties = 0;
  let leftPicks = 0;
  for (const ann of Object.values(effective)) {
    const winner = ann.winner ?? ann.choice;
    if (winner === "a") aWins += 1;
    else if (winner === "b") bWins += 1;
    else if (winner === "tie") ties += 1;
    if ((winner === "a" || winner === "b") && ann.shown_left) leftPicks += Number(winner === ann.shown_left);
  }
  const decided = aWins + bWins;
  return {
    aWins,
    bWins,
    ties,
    decided,
    winRate: decided ? bWins / decided : null,
    ci: wilsonInterval(bWins, decided),
    pValue: decided ? signTestP(bWins, decided) : null,
    leftShare: decided ? leftPicks / decided : null
  };
}

// (judge recommendation, human label) pairs for every reviewed example with a verdict.
export function judgeHumanPairs(effective: Record<string, Annotation>, verdicts: Record<string, Verdict>): Array<[string, string]> {
  const pairs: Array<[string, string]> = [];
  for (const ann of Object.values(effective)) {
    if (ann.label !== "p" && ann.label !== "f") continue;
    const verdict = verdicts[ann.example_id];
    if (verdict) pairs.push([verdict.recommendation, ann.label]);
  }
  return pairs;
}
